backend/runner/controllerproxy.go

```diff
@@ -0,0 +1,114 @@
+package runner
+
+import (
+	"context"
+	"fmt"
+	"net/url"
+
+	"connectrpc.com/connect"
+
+	ftlv1 "github.com/TBD54566975/ftl/backend/protos/xyz/block/ftl/v1"
+	"github.com/TBD54566975/ftl/backend/protos/xyz/block/ftl/v1/ftlv1connect"
+	"github.com/TBD54566975/ftl/internal/log"
+	"github.com/TBD54566975/ftl/internal/rpc"
+)
+
+var _ ftlv1connect.VerbServiceHandler = &runnerProxy{}
+var _ ftlv1connect.ModuleServiceHandler = &runnerProxy{}
+
+type runnerProxy struct {
+	controllerVerbService   ftlv1connect.VerbServiceClient
+	controllerModuleService ftlv1connect.ModuleServiceClient
+	bindAddress             *url.URL
+}
+
+func runProxyServer(ctx context.Context, controllerVerbService ftlv1connect.VerbServiceClient, controllerModuleService ftlv1connect.ModuleServiceClient) (*runnerProxy, error) {
+	proxy := &runnerProxy{
+		controllerVerbService:   controllerVerbService,
+		controllerModuleService: controllerModuleService,
+	}
+	logger := log.FromContext(ctx)
+	parse, err := url.Parse("http://127.0.0.1:0")
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse url: %w", err)
+	}
+	proxyServer, err := rpc.NewServer(ctx, parse,
+		rpc.GRPC(ftlv1connect.NewVerbServiceHandler, proxy),
+		rpc.GRPC(ftlv1connect.NewModuleServiceHandler, proxy),
+	)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create server: %w", err)
+	}
+	urls := proxyServer.Bind.Subscribe(nil)
+	go func() {
+		err := proxyServer.Serve(ctx)
+		if err != nil {
+			logger.Errorf(err, "failed to serve")
+			return
+		}
+	}()
+	proxy.bindAddress = <-urls
+	return proxy, nil
+}
+
+func (r *runnerProxy) GetModuleContext(ctx context.Context, c *connect.Request[ftlv1.GetModuleContextRequest], c2 *connect.ServerStream[ftlv1.GetModuleContextResponse]) error {
+	moduleContext, err := r.controllerModuleService.GetModuleContext(ctx, connect.NewRequest(c.Msg))
+	if err != nil {
+		return fmt.Errorf("failed to get module context: %w", err)
+	}
+	for {
+		rcv := moduleContext.Receive()
+		if rcv {
+			err := c2.Send(moduleContext.Msg())
+			if err != nil {
+				return fmt.Errorf("failed to send message: %w", err)
+			}
+		} else {
```

Comment:
There should be a check here for if `moduleContext.Err() != nil`, which can happen if the server closes the connection cleanly. Otherwise below will report a nil error as an error.